sdks/python/apache_beam/runners/worker/worker_pool_main.py

```diff
@@ -51,6 +51,28 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+def kill_process_gracefully(proc, timeout=10):
+  """
+  Kill a worker process gracefully by sending a SIGTERM and waiting for
+  it to finish. A SIGKILL will be sent if the process has not finished
+  after ``timeout`` seconds.
+  """
+  def _kill():
+    proc.terminate()
+    t = time.time()
+    while time.time() < t + timeout:
+      time.sleep(0.01)
+      if proc.poll() is not None:
```

Comment:
why not proc.wait() directly without looping?

Reply:
Because wait blocks. The loop is supposed to exit if the process exits or after 5 seconds. The latter won't happen with `wait()`.

Reply:
you can proc.wait(timeout=timeout), isn't it?

Reply:
True, but that wouldn't be less code, because it requires a surrounding try/except block. The docs also suggest that wait with timeout does busy wait, so doing an explicit while with sleep is probably better anyway.

Reply:
I changed it to wait with timeout, but again, we can reduce it even more if we can guarantee that the child process will not hang indefinitly.